database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/PostViewModel.swift

```diff
@@ -0,0 +1,123 @@
+//
+//  Copyright (c) 2021 Google Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import SwiftUI
+import Firebase
+
+class PostViewModel: ObservableObject, Identifiable {
+  @Published var id: String
+  @Published var uid: String
+  @Published var author: String
+  @Published var title: String
+  @Published var body: String
+  @Published var starCount: Int
+  @Published var userIDsStarredBy: [String: Bool]
+  var postRef: DatabaseReference!
+
+  init(id: String, uid: String, author: String, title: String, body: String) {
+    self.id = id
+    self.uid = uid
+    self.author = author
+    self.title = title
+    self.body = body
+    starCount = 0
+    userIDsStarredBy = [:]
+  }
+
+  init?(id: String, dict: [String: Any]) {
+    guard let uid = dict["uid"] as? String else { return nil }
+    guard let author = dict["author"] as? String else { return nil }
+    guard let title = dict["title"] as? String else { return nil }
+    guard let body = dict["body"] as? String else { return nil }
+    let userIDsStarredBy = dict["userIDsStarredBy"] as? [String: Bool] ?? [:]
+    let starCount = dict["starCount"] as? Int ?? 0
+
+    self.id = id
+    self.uid = uid
+    self.author = author
+    self.title = title
+    self.body = body
+    self.starCount = starCount
+    self.userIDsStarredBy = userIDsStarredBy
+  }
+
+  func isStarred() -> Bool {
+    if let uid = Auth.auth().currentUser?.uid {
+      return userIDsStarredBy[uid] ?? false
```

Comment:
nit: I think we can improve implementation because the current one has a couple issues:
1. If `Auth.auth().currentUser?.uid` updates so that `isStarred` value changes, the view will not be updated
2. property `userIDsStarredBy` is an implementation detail and should not be exposed to the view but currently it's the one that triggers view re-asking `isStarred` when needed.

It can be done in a separate PR.